### DISCUSSION TITLE
Viseron for Jetson Nano on docker lumaia/viseron-jetson

### DISCUSSION THREAD

**Discussion by peterek102** [roflcoopter/viseron#159] [General]
Hi,

https://hub.docker.com/r/lumaia/viseron-jetson

Could someone identify as the owner of this project? Is it available on Github or elsewhere? I tried to run it on Jetson Nano but with no luck.

Traceback (most recent call last):
  File "viseron.py", line 6, in <module>
    from const import LOG_LEVELS
  File "/src/viseron/const.py", line 1, in <module>
    import cv2
  File "/usr/lib/python3.6/dist-packages/cv2/__init__.py", line 102, in <module>
    bootstrap()
  File "/usr/lib/python3.6/dist-packages/cv2/__init__.py", line 90, in bootstrap
    import cv2
ImportError: libnppc.so.10: cannot open shared object file: No such file or directory

@roflcoopter  I own Jeston-Nano and RPI 4 and have some spare time. If you still need to test on hardware let me know.

BR
Pit

**Comment by roflcoopter** [comment]
Hi! Its discussed a little in this issue, #108.

I would love if you could try out the `roflcoopter/viseron:1.8.0b6` on your RPi 4.
There are still some bugs that i need help with since i dont own an RPi4.

My plan is to begin writing the Jetson NANO version after the RPi4 one is done.

Would very much appreciate your help!

**Reply by peterek102** [reply]
@roflcoopter should I test it on 32 or 64 bit version ?

**Reply by roflcoopter** [reply]
64-bit, its a multiarch image and the real underlying image is `roflcoopter/aarch64-viseron:1.8.0b6`

**Reply by peterek102** [reply]
No luck with starting it. It tried to run it both as roflcoopter/viseron:1.8.0b6 and directly roflcoopter/aarch64-viseron:1.8.0b6 but received the same warning after which it closes. Maybe I'm doing something wrong trying to test if it starts correctly? Are there any requirements for the host?

WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested

Full log:
pi@raspberrypi:~/Documents/Docker $ sudo docker container run -it roflcoopter/aarch64-viseron:1.8.0b6
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing...
usermod: no changes
************************ UID/GID *************************
User uid:    911
User gid:    911
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing...
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing...
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] 40-gid-vaapi: executing...
[cont-init.d] 40-gid-vaapi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Unable to find configuration. Creating default one in /config/config.yaml
Please fill in the necessary configuration options and restart Viseron
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.

` @Docker installation log:`
Client: Docker Engine - Community
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        370c289
 Built:             Fri Apr  9 22:46:02 2021
 OS/Arch:           linux/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:44:13 2021
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

**Comment by roflcoopter** [comment]
What happens if you add `--platform linux/arm64/v8` to your command?

It might be a bug with Docker itself, see this [discussion](https://github.com/dani-garcia/bitwarden_rs/discussions/1264)

**Reply by peterek102** [reply]
Still no luck
@roflcoopter I can give you access to this device over OpenVPN.

full log
pi@raspberrypi:~ $ sudo docker container run -it --platform linux/arm64/v8 roflcoop
ter/viseron:1.8.0b6
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing...
usermod: no changes
************************ UID/GID *************************
User uid:    911
User gid:    911
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing...
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing...
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] 40-gid-vaapi: executing...
[cont-init.d] 40-gid-vaapi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Unable to find configuration. Creating default one in /config/config.yaml
Please fill in the necessary configuration options and restart Viseron
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.

**Reply by roflcoopter** [reply]
You need to mount a config folder to `/config`, so add `-v <config path>:/config` to your Docker run command

**Reply by peterek102** [reply]
@roflcoopter OK.One step further. I haven't configured any camera in config.yaml yet.

pi@raspberrypi:~/Docker/config $ sudo docker container run -it -v /home/pi/Docker/config:/config --platform linux/arm64/v8 roflcoopter/viseron:1.8.0b6
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing...
usermod: no changes
************************ UID/GID *************************
User uid:    911
User gid:    911
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing...
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing...
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] 40-gid-vaapi: executing...
[cont-init.d] 40-gid-vaapi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/lib/python3.7/runpy.py", line 142, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/usr/lib/python3.7/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/src/viseron/__init__.py", line 9, in <module>
    from viseron.config import CONFIG, NVRConfig, ViseronConfig
  File "/src/viseron/config/__init__.py", line 95, in <module>
    CONFIG = VISERON_CONFIG_SCHEMA(raw_config)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected int for dictionary value @ data['cameras'][0]['port']
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.

**Reply by roflcoopter** [reply]
You have to configure atleast one camera for Viseron to start properly

**Reply by peterek102** [reply]
I configured 1 camera (rtsp stream) - configuration tested on VLC (working). I only configured what was in the default config.yaml.  I modified docker command to match readme. What is the minimum set to make viseron work as there is still error in the log?

Configuration

```

# See the README for the full list of configuration options.
cameras:
  - name: test
    host: 10.178.0.148
    port: 554
    username: ***
    password: ***
    path: /cam/realmonitor?channel=6&subtype=0&unicast=true&proto=Onvif

# MQTT is optional
mqtt:
  broker: maqiatto.com
  port: 1883
  username: ***
  password: ***
```

Log 

pi@raspberrypi:~/Docker/viseron/config3 $ sudo docker container run --rm --privileged -v /home/pi/Docker/viseron/config3:/config --platform linux/arm64/v8 roflcoopter/viseron:1.8.0b6
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing...
usermod: no changes
************************ UID/GID *************************
User uid:    911
User gid:    911
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing...
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing...
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] 40-gid-vaapi: executing...
[cont-init.d] 40-gid-vaapi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/lib/python3.7/runpy.py", line 142, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/usr/lib/python3.7/runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "/src/viseron/__init__.py", line 9, in <module>
    from viseron.config import CONFIG, NVRConfig, ViseronConfig
  File "/src/viseron/config/__init__.py", line 95, in <module>
    CONFIG = VISERON_CONFIG_SCHEMA(raw_config)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected a list for dictionary value @ data['cameras']
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.

**Reply by roflcoopter** [reply]
That should be enough. I just tested with a minimal configf and it works.
Are you sure your config file is mounted properly into the container? Can you paste your docker run command?

**Reply by peterek102** [reply]
@roflcoopter 

`sudo docker container run --rm --privileged -v /home/pi/Docker/viseron/config3:/config --platform linux/arm64/v8 roflcoopter/viseron:1.8.0b6`

It's also in the log above

**Reply by roflcoopter** [reply]
whats the output of `sudo docker container run --rm --privileged --entrypoint /bin/bash -v /home/pi/Docker/viseron/config3:/config --platform linux/arm64/v8 roflcoopter/viseron:1.8.0b6 cat /config/config.yaml`

**Reply by peterek102** [reply]
I made a clean rasbian install in the meantime and now it started. Probably some Python related libraries installed on the host caused the issue. 
I had connection issue with MQTT broker (maqiatto.com) so I disabled in in the config.

`sudo docker container run --rm --privileged -v /home/pi/Docker/viseron/config3:/config --platform linux/arm64/v8 roflcoopter/viseron:1.8.0b6`

[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing...
usermod: no changes
************************ UID/GID *************************
User uid:    911
User gid:    911
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing...
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing...
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] 40-gid-vaapi: executing...
[cont-init.d] 40-gid-vaapi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[2021-04-30 22:54:58] [viseron                 ] [INFO    ] - -------------------------------------------
[2021-04-30 22:54:58] [viseron                 ] [INFO    ] - Initializing...
[2021-04-30 22:54:59] [viseron.detector.edgetpu] [ERROR   ] - EdgeTPU not found. Detection will run on CPU
[2021-04-30 22:54:59] [viseron                 ] [INFO    ] - Initializing NVR threads
[2021-04-30 22:55:02] [viseron                 ] [INFO    ] - Initialization complete

@roflcoopter do you have any test procedure you would like to run at this point? I think it would be best to check if GPU is utilized. Looking at the log it seems like CPU is going to be used.

**Reply by roflcoopter** [reply]
Sorry i have been busy the last few days!
You need to mount some more volumes for the EdgeTPU to work, try to add `-v /dev/bus/usb:/dev/bus/usb` to your docker run command.

Also it would be good if you can add debug logging to config.yaml, like this
```yaml
logging:
  level: debug
```

**Comment by e1z0** [comment]
```
[2021-10-11 17:32:37] [DEBUG   ] [viseron.nvr.front_camera_1] - Waiting for first frame
[2021-10-11 17:32:37] [DEBUG   ] [viseron.nvr.front_camera_1] - NVR thread initialized
[2021-10-11 17:32:37] [INFO    ] [viseron                 ] - Initialization complete
[2021-10-11 17:32:37] [DEBUG   ] [viseron.mqtt            ] - Got topic viseron/viseron_front_door/switch/set, message ON
[2021-10-11 17:32:39] [ERROR   ] [viseron.camera.stream.front_camera_1] - NvMMLiteOpen : Block : BlockType = 261 
[2021-10-11 17:32:39] [ERROR   ] [viseron.camera.stream.front_camera_1] - NVMEDIA: Reading vendor.tegra.display-size : status: 6 
[2021-10-11 17:32:39] [ERROR   ] [viseron.camera.stream.front_camera_1] - NvMMLiteBlockCreate : Block : BlockType = 261 
[2021-10-11 17:33:49] [DEBUG   ] [viseron.watchdog.thread_watchdog] - Thread viseron.camera.front_camera_1 is stuck
```
